odl/tomo/data/mrc2014.py

```diff
@@ -0,0 +1,846 @@
+            # Need to invert the permutation
+            return tuple(ncells[ax]
+                         for ax in np.argsort(self.data_axis_order))
+
+    @property
+    def data_dtype(self):
+        """Data type of the data block as determined from `header`.
+
+        If the header entry ``'mode'`` is missing, ``None`` is returned.
+        Otherwise, it is determined from ``mode``.
+        """
+        if not self.header:
+            return None
+        try:
+            mode = int(self.header['mode']['value'])
+        except KeyError:
+            return None
+        else:
+            try:
+                return MRC_MODE_TO_NPY_DTYPE[mode]
+            except KeyError:
+                raise ValueError('data mode {} not supported'.format(mode))
+
+    @property
+    def data_axis_order(self):
+        """Permutation of ``(0, 1, 2)``.
+
+        If one of the header entries ``'mapc', 'mapr', 'maps'`` is missing,
+        the identity permutation ``(1, 2, 3)`` is returned.
+        Otherwise, value is determined from the ``'mapc', 'mapr', 'maps'``
+        header entries.
+        """
+        if not self.header:
+            return (0, 1, 2)
+        try:
+            mapc = self.header['mapc']['value']
+            mapr = self.header['mapr']['value']
+            maps = self.header['maps']['value']
+        except KeyError:
+            return (0, 1, 2)
+        else:
+            return tuple(int(m) - 1 for m in (mapc, mapr, maps))
+
+    def write_data(self, data, dstart=None, swap_axes=True):
+        """Write ``data`` to `file`.
+
+        Parameters
+        ----------
+        data : `array-like`
+            Data that should be written to `file`.
+        dstart : non-negative int, optional
+            Offset in bytes of the start position of the written data.
+            If provided, reshaping and axis swapping of ``data`` is
+            skipped.
+            For ``None``, `header_size` is used.
+        swap_axes : bool, optional
+            If ``True``, use the ``'mapc', 'mapr', 'maps'`` header entries
+            to swap the axes in the ``data`` before writing. Use ``False``
+            only if the data is already consistent with the final axis
+            order.
+        """
+        if dstart is None:
+            shape = self.data_shape
+            dstart = int(self.header_size)
+        elif dstart < 0:
+            raise ValueError('`dstart` must be non-negative, got {}'
+                             ''.format(dstart))
+        else:
+            shape = -1
+            dstart = int(dstart)
+
+        if dstart < self.header_size:
+            raise ValueError('invalid `dstart`, resulting in absolute '
+                             '`dstart` < `header_size` ({} < {})'
+                             ''.format(dstart, self.header_size))
+
+        data = np.asarray(data, dtype=self.data_dtype).reshape(shape)
+        if swap_axes:
+            data = np.transpose(data, axes=self.data_axis_order)
```

Comment:
Fixed this now, but not very much has changed in the end. I had to get a better idea of the relation between `axis_order` and `(mapc, mapr, maps)`. I understand the two concept as follows:

- `axis_order[1] = 0` : the data axis 0 corresponds to storage axis 1 
- `mapr = 0` : rows map to axis 0, i.e. storage axis 1 corresponds to data axis 0

So they're the same. Perhaps one should call it `storage_axis_order` for clarity.

Then, there is a necessary asymmetry between reader and writer. Reader needs to swap from storage shape to data shape, writer the other way around. Now the `np.transpose` function needs as `axes` argument the permutation "which axis goes where", i.e. the result of an axis permutation *to* the desired result. Since `axis_order` maps *storage order to data order*, we can use it directly in the reader and its inverse for the writer.